zebra-chain/src/commitments/sapling.rs

```diff
@@ -0,0 +1,233 @@
+//! Sapling note and value commitments and types.
+
+#[cfg(test)]
+mod arbitrary;
+#[cfg(test)]
+mod test_vectors;
+
+pub mod pedersen_hashes;
+
+use std::{convert::TryFrom, fmt, io};
+
+use bitvec::prelude::*;
+use rand_core::{CryptoRng, RngCore};
+
+use crate::{
+    keys::sapling::{find_group_hash, Diversifier, TransmissionKey},
+    serde_helpers,
+    serialization::{ReadZcashExt, SerializationError, ZcashDeserialize, ZcashSerialize},
+    types::amount::{Amount, NonNegative},
+};
+
+use pedersen_hashes::*;
+
+/// The randomness used in the Pedersen Hash for note commitment.
+#[derive(Copy, Clone, Debug, PartialEq)]
+pub struct CommitmentRandomness(jubjub::Fr);
+
+/// Note commitments for the output notes.
+#[derive(Clone, Copy, Deserialize, PartialEq, Serialize)]
+pub struct NoteCommitment(#[serde(with = "serde_helpers::AffinePoint")] pub jubjub::AffinePoint);
```

Comment:
I don't think we should use an `AffinePoint` here, since it's expensive to construct and we won't always want to process the decompressed value.  (Not sure where the previous discussion went, I tried to follow the link and it seemed to have broken, but that could just be my error).

Reply:
Discussed offline that we are going to have the internal representation of these point types as the canonical bytes encoding. This means that when parsing a `Transaction`, say one we receive gossip'd peer to peer, a `Transaction` with an off-curve point as an ephemeral key or commitment, will not be checked until we decode and check those values in order to validate the enclosing `Block` or as part of mining if we write a mempool validator. We're ok with this, as we're probably only going to check all the consensus rules once anyway, and storing as bytes will save us scalar muls every time we deserialize these types.

Reply:
UPDATE: deferring this to later, lots of TryFroms and Results downstream